### PR TITLE
Feat/edge identity viewset

### PR DESCRIPTION
**Pull request by gagantrivedi**
Add view set for edge identities.
hosted at: `/environments/{environment_api_key}/edge-identities/`

### CODE REVIEW

File: api/projects/serializers.py [hunk 1]
```diff
@@ -15,7 +15,13 @@
 class ProjectSerializer(serializers.ModelSerializer):
     class Meta:
         model = Project
-        fields = ("id", "name", "organisation", "hide_disabled_flags")
+        fields = (
+            "id",
+            "name",
+            "organisation",
+            "hide_disabled_flags",
+            "enable_dynamo_db",
```
Comment (gagantrivedi) [inline api/projects/serializers.py:23]: Should I move this to a different PR?

File: api/app/pagination.py [hunk 2]
```diff
@@ -1,7 +1,61 @@
...
+    def paginate_queryset(self, dynamo_queryset, request, view=None):
+        self.previous_last_evaluated_key = request.GET.get("last_evaluated_key")
+        self.request = request
+        last_evaluated_key = dynamo_queryset.get("LastEvaluatedKey")
+        if last_evaluated_key:
+            self.last_evaluated_key = base64.b64encode(
+                json.dumps(last_evaluated_key).encode()
+            )
+
+        return [
```
Comment (gagantrivedi) [inline api/app/pagination.py:27]: I don't feel very good about converting dynamo document to model here what do you think? but if we do that on the `get_queryset` method than we have to override list to pass `last_evaluated_key` to this

Comment (matthewelwell) [reply api/app/pagination.py:27]: >I don't feel very good about converting dynamo document to dict

I think you mean dynamo document to engine model? 

>but if we do that on the get_queryset method than we have to override list to pass last_evaluated_key to this

I don't really follow this statement. Can you illustrate your point with a code snippet perhaps?

File: api/environments/identities/models.py [hunk 3]
```diff
@@ -19,6 +21,72 @@
         settings.IDENTITIES_TABLE_NAME_DYNAMO
     )
 
+dynamo_identifier_search_functions = {
+    "EQUAL": lambda identifier: Key("identifier").eq(identifier),
+    "BEGINS_WITH": lambda identifier: Key("identifier").begins_with(identifier),
+}
+
+
+class DynamoIdentityManager(models.Manager):
```
Comment (matthewelwell) [inline api/environments/identities/models.py]: This class should live in a `managers` module.

File: api/environments/identities/models.py [hunk 4]
```diff
@@ -27,6 +95,8 @@ class Identity(models.Model):
     environment = models.ForeignKey(
         Environment, related_name="identities", on_delete=models.CASCADE
     )
+    objects = models.Manager()
```
Comment (matthewelwell) [inline api/environments/identities/models.py]: This is unnecessary - it's the default.

File: api/environments/identities/models.py (hunk 3 above)
Comment (matthewelwell) [reply api/environments/identities/models.py]: In fact, does it even need to override the Manager class? What are we using from the `Manager` class?

File: api/projects/serializers.py (hunk 1 above)
Comment (matthewelwell) [reply api/projects/serializers.py:23]: No, I'd say it's fine here.

File: api/environments/identities/views.py [hunk 5]
```diff
@@ -21,6 +30,62 @@
...
+
+    @staticmethod
+    def _get_search_function_and_value(
+        search_query: str,
+    ) -> typing.Tuple[typing.Callable, str]:
+        if search_query.startswith('"') and search_query.endswith('"'):
+            return dynamo_identifier_search_functions["EQUAL"], search_query.replace(
+                '"', ""
+            )
+        else:
```
Comment (matthewelwell) [inline api/environments/identities/views.py]: This `else` is unnecessary. Remember that the pythonic way of doing things is to always have a 'default' return.

File: api/environments/identities/views.py [hunk 6]
```diff
@@ -21,6 +30,62 @@
...
+        identity_documents = Identity.dynamodb.search_items_with_identifier(
+            self.kwargs["environment_api_key"],
+            search_identifier,
+            search_func,
+            page_size,
+            start_key,
+        )
+        return identity_documents
+
+    def get_environment_from_request(self):
```
Comment (matthewelwell) [inline api/environments/identities/views.py:88]: Why do we need this?

File: api/environments/identities/views.py [hunk 7]
```diff
@@ -21,6 +30,62 @@
 from util.views import SDKAPIView
 
 
+class EdgeIdentityViewSet(viewsets.ModelViewSet):
+    serializer_class = EdgeIdentitySerializer
+    permission_classes = [IsAuthenticated, NestedEnvironmentPermissions]
+    pagination_class = EdgeIdentityPagination
+    lookup_field = "identity_uuid"
+
+    @staticmethod
```
Comment (matthewelwell) [inline api/environments/identities/views.py]: I don't really like seeing static methods that are only marked as static because they don't use anything from the class. Static methods should be used outside of the class like `Class.static_method()`. Could / should we move the `dynamo_identifier_search_functions` constant into this class?

File: api/environments/identities/views.py (hunk 6 above)
Comment (gagantrivedi) [reply api/environments/identities/views.py:88]: I guess we can inline this?

Comment (matthewelwell) [reply api/environments/identities/views.py:88]: But it's not even used is it?

Comment (gagantrivedi) [reply api/environments/identities/views.py:88]: Oh it is being used here: https://github.com/Flagsmith/flagsmith/pull/530/files#diff-55e28274d29ce084198b7aacadcd41d63b764f7adea2d1be9d6b15f19da96907R41. It was not being used in the initial commit tho, so that's my bad

File: api/environments/identities/views.py (hunk 7 above)
Comment (gagantrivedi) [reply api/environments/identities/views.py]: hmm.. yeah, I was trying to hide internal filters of dynamodb from view layer, but I guess it's okay since we do that for django models as well.

File: api/app/pagination.py (hunk 2 above)
Comment (gagantrivedi) [reply api/app/pagination.py:27]: >I think you mean dynamo document to engine model?

yeah that is correct 

>I don't really follow this statement. Can you illustrate your point with a code snippet perhaps?

yeah, so currently we are using the default implementation of `list`, because `get_queryset` returns the raw response of `dynamodb query api` instead of returning a list of engine model instances. But if we update `get_queryset` to return a list of engine model instance instead of the raw dynamodb response, we will have to override list to look something like

```
    def list(self, request, *args, **kwargs):
        queryset = self.filter_queryset(self.get_queryset())
        last_evaluated_key = some_method()
        page = self.paginate_queryset(queryset, last_evaluated_key)
        if page is not None:
            serializer = self.get_serializer(page, many=True)
            return self.get_paginated_response(serializer.data)

        serializer = self.get_serializer(queryset, many=True)
        return Response(serializer.data)



```

Comment (matthewelwell) [reply api/app/pagination.py:27]: Ok, understood. Then yeah, I definitely prefer this approach.

File: api/environments/identities/models.py [hunk 8]
```diff
@@ -20,13 +22,75 @@
     )
 
 
+class DynamoIdentityManager:
```
Comment (matthewelwell) [inline api/environments/identities/models.py]: Can we extract all of this into a separate `dynamo` module? We should write some kind of wrapper for all of the dynamo logic there and then just instantiate the wrapper here. That goes for the writing to dynamo piece as well - it should all be done through the wrapper not this pseudo manager class.

**Review by matthewelwell — CHANGES_REQUESTED** [review]

File: api/environments/dynamodb/dynamodb_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
+import typing
+
+import boto3
+from boto3.dynamodb.conditions import Key
+from django.conf import settings
+from django.core.exceptions import ObjectDoesNotExist
+
+
+class DynamoIdentityWrapper:
+    _table = None
```
Comment (matthewelwell) [inline api/environments/dynamodb/dynamodb_wrapper.py]: I think this should be done in `__init__`. I think this could cause some strange behaviour otherwise.

File: api/environments/identities/models.py [hunk 10]
```diff
@@ -1,23 +1,17 @@
...
 from features.multivariate.models import MultivariateFeatureStateValue
 
-# Intialize the dynamo client globally
-dynamo_identity_table = None
-if settings.IDENTITIES_TABLE_NAME_DYNAMO:
-    dynamo_identity_table = boto3.resource("dynamodb").Table(
-        settings.IDENTITIES_TABLE_NAME_DYNAMO
-    )
+# Initialize the dynamo wrapper client globally
+dynamo_identity_wrapper = DynamoIdentityWrapper()
```
Comment (matthewelwell) [inline api/environments/identities/models.py]: Why initialise this here only to import it in `serializers`? We should either initialise it in the `serializers` module or add the logic back to the model class here (just not using the manager). We could use class methods or similar?

File: api/environments/identities/views.py [hunk 11]
```diff
@@ -1,15 +1,22 @@
...
 from rest_framework.permissions import IsAuthenticated
 from rest_framework.response import Response
 from rest_framework.schemas import AutoSchema
 
-from app.pagination import CustomPagination
+from app.pagination import CustomPagination, EdgeIdentityPagination
 from environments.identities.helpers import identify_integrations
-from environments.identities.models import Identity
-from environments.identities.serializers import IdentitySerializer
+from environments.identities.models import Identity, dynamo_identity_wrapper
```
Comment (matthewelwell) [inline api/environments/identities/views.py]: Yeah, this isn't right, we shouldn't be passing the `dynamo_identity_wrapper` object around between modules like this. Let's move all the logic that uses the wrapper into the `models` module using methods on the `Identity` class (staticmethod, classmethod or otherwise)

**Review by matthewelwell — CHANGES_REQUESTED** [review]

File: api/environments/identities/views.py (hunk 11 above)
Comment (gagantrivedi) [reply api/environments/identities/views.py]: something like
```

@python_2_unicode_compatible
class Identity(models.Model):
    identifier = models.CharField(max_length=2000)
    created_date = models.DateTimeField("DateCreated", auto_now_add=True)
    environment = models.ForeignKey(
        Environment, related_name="identities", on_delete=models.CASCADE
    )

    dynamo_identity_wrapper = DynamoIdentityWrapper()



```

which makes the interface looks something like

```
        return Identity.dynamo_identity_wrapper.get_item_from_uuid(
            self.kwargs["environment_api_key"], self.kwargs["identity_uuid"]
        )



```

File: api/environments/dynamodb/dynamodb_wrapper.py (hunk 9 above)
Comment (gagantrivedi) [reply api/environments/dynamodb/dynamodb_wrapper.py]: ah that can make mocking a bit more pain

Comment (matthewelwell) [reply api/environments/dynamodb/dynamodb_wrapper.py]: I'd rather the mocking was a pain than the actual production logic... As far as I'm aware, you shouldn't really set anything mutable as an attribute directly on the class. Couldn't you extract the logic to a `get_dynamo_table` function that you can subsequently mock?

**Review by matthewelwell — APPROVED** [review]
